nrf52-hal-common/src/uarte.rs

```diff
@@ -0,0 +1,153 @@
+        uarte.psel.txd.write(|w| {
+            let w = unsafe { w.pin().bits(pins.txd.pin) };
+            w.connect().connected()
+        });
+
+        // Optional pins
+        uarte.psel.cts.write(|w| {
+            if let Some(ref pin) = pins.cts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        uarte.psel.rts.write(|w| {
+            if let Some(ref pin) = pins.rts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        // Enable UARTE instance
+        uarte.enable.write(|w| w.enable().enabled());
+
+        // Configure
+        uarte.config.write(|w| {
+            w.hwfc()
+                .bit(pins.rts.is_some() && pins.cts.is_some())
+                .parity()
+                .variant(parity)
+        });
+
+        // Configure frequency
+        uarte.baudrate.write(|w| w.baudrate().variant(baudrate));
+
+        Uarte(uarte)
+    }
+
+    /// Write via UART
+    ///
+    /// This method uses transmits all bytes in `tx_buffer`
+    ///
+    /// The buffer must have a length of at most 255 bytes
+    pub fn write(&mut self, tx_buffer: &[u8]) -> Result<(), Error> {
+        // This is overly restrictive. See (similar SPIM issue):
+        // https://github.com/nrf-rs/nrf52/issues/17
+        if tx_buffer.len() > u8::max_value() as usize {
+            return Err(Error::TxBufferTooLong);
+        }
+
+        // Set up the DMA write
+        self.0.txd.ptr.write(|w|
+            // We're giving the register a pointer to the stack. Since we're
+            // waiting for the UART transaction to end before this stack pointer
+            // becomes invalid, there's nothing wrong here.
+            //
+            // The PTR field is a full 32 bits wide and accepts the full range
+            // of values.
+            unsafe { w.ptr().bits(tx_buffer.as_ptr() as u32) });
+        self.0.txd.maxcnt.write(|w|
+            // We're giving it the length of the buffer, so no danger of
+            // accessing invalid memory. We have verified that the length of the
+            // buffer fits in an `u8`, so the cast to `u8` is also fine.
+            //
+            // The MAXCNT field is 8 bits wide and accepts the full range of
+            // values.
+            unsafe { w.maxcnt().bits(tx_buffer.len() as _) });
+
+        // Start UART Receive transaction
+        self.0.tasks_starttx.write(|w|
+            // `1` is a valid value to write to task registers.
+            unsafe { w.bits(1) });
+
+        // Wait for transmission to end
+        while self.0.events_endtx.read().bits() == 0 {}
+
+        // Reset the event, otherwise it will always read `1` from now on.
+        self.0.events_endtx.write(|w| w);
```

Comment:
Same as the previous comment: We should create a generic event API, so all peripheral code doesn't have to remember to reset event flags and such.

The regularity and well-thought-out-ed-ness of the Nordic peripheral APIs lends itself really well to a mid-level API layer, between the low-level svd2rust API and the high-level HAL API.

Reply:
I was thinking about that, the interesting part will be handling different cases around which/how many events to set for each peripheral. For example, SPIM has one for tx+rx, while UARTE has one for each of tx and rx, as they are independent. I'm sure we can find something though.

Reply:
Here, too, I was thinking about something less ambitious. Something like that:
``` rust
impl Event {
    pub fn peek(&self) -> bool {
        // Return bit from register
    }

    pub fn consume(&mut self) -> bool {
        // Return bit from register and reset it
    }
}
```
And have a macro that generates that for every `EVENTS_` register, to cut down on duplicated code.